File: 리트코드/0147-insertion-sort-list/0147-insertion-sort-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        
        ListNode parent = new ListNode();
        ListNode p = parent; 
        while(head != null){
               
            while(p.next != null && p.next.val < head.val){
                p = p.next;
            }

            ListNode pNext = p.next;
            ListNode hNext = head.next;
            p.next = head;
            head.next = pNext;
            head = hNext;

            if(head != null && p.next.val > head.val){
                p = parent;
            }
        }

        return parent.next;
    }
}

File: 리트코드/0147-insertion-sort-list/0147-insertion-sort-list.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun insertionSortList(head: ListNode?): ListNode? {

        var cur:ListNode? = head
        val parent:ListNode = ListNode()
        var p:ListNode = parent
        while(cur != null){
            while(p.next != null && p.next.`val` < cur.`val`){
                p = p.next
            }

            val pNext:ListNode? = p.next
            val cNext:ListNode? = cur.next

            p.next = cur
            cur.next = pNext
            cur = cNext

            if(cur != null && p.next.`val` > cur.`val`){
                p = parent
            }
        }

        return parent.next
    }
}